### ISSUE TITLE
How to use a different column name for JSON serialization and deserialization in GRDB?

### ISSUE THREAD

**Issue by SF-Simon** [groue/GRDB.swift#1675]
Hello, @groue  we have a special issue. We need to assign specific keys to certain fields in the CodingKeys. We're a bit unclear about how this assignment works. Please take a look at the class in this demonstration.

```
class Player: FetchableRecord & PersistableRecord, Codable {
    var id: Int64
    var name: String
    var score: Int

    /// JSON Keys
    enum CodingKeys: String, CodingKey {
        case id
        case name = "title"  // <<<----- here is key
        case score
    }
}
```

I would like to ask whether setting the `name = "title"` key will affect the creation and reading of fields in GRDB. In other words, will the name variable's field `name` in SQLite3 be changed due to this setting? 

In other words, without defining the `enum Columns {}` explicitly, will the `enum CodingKeys {}` affect the table's column names?

**Comment by SF-Simon** [comment]
We would like to use "Title" as the key during JSON serialization and deserialization, but in the GRDB database table, the field is named "name." Is there a way to achieve this in GRDB? If so, how can it be implemented?

**Comment by groue** [comment]
Hello @SF-Simon,

As you might have seen by running your code, or reading the [Codable documentation](https://developer.apple.com/documentation/foundation/archives_and_serialization/encoding_and_decoding_custom_types), the `CodingKeys` enum drives the keys of an encoded player in all encoders (JSON, database, etc).

> We would like to use "Title" as the key during JSON serialization and deserialization, but in the GRDB database table, the field is named "name." Is there a way to achieve this in GRDB? If so, how can it be implemented?

My favorite solution is to define two distinct `Codable` types, one for the database, and one for JSON. It allows JSON and database representations to evolve independently, with minimal fuss. Converting between those two types may be verbose, but it is not complicated.

Other possible solutions are more involved:

1. Harmonize database and JSON representations (rename the "name" column to "title"). See [Defining the Database Schema from a Migration](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/migrations#Defining-the-Database-Schema-from-a-Migration) for some guidance.

2. Keep a single `Player` type, which uses `Codable` for JSON (it's not like we have a choice), and opts out of `Codable` for database persistence:

    ```swift
    struct Player {
        var id: Int64
        var name: String
        var score: Int
    }

    // JSON support
    extension Player: Codable {
        enum CodingKeys: String, CodingKey {
            case id
            case name = "title"
            case score
        }
    }

    // Database support
    extension Player: FetchableRecord, PersistableRecord {
        enum Columns {
            static let id = Column(CodingKeys.id)
            static let name = Column("name") // deviation from JSON
            static let score = Column(CodingKeys.score)
        }
        
        init(row: Row) {
            self.init(
                id: row[Columns.id],
                name: row[Columns.name],
                score: row[Columns.score])
        }
        
        func encode(to container: inout PersistenceContainer) {
            container[Columns.id] = id
            container[Columns.name] = name
            container[Columns.score] = score
        }
    }
    ```

**Comment by SF-Simon** [comment]
Thank you @groue , I’ve learned something new. 

However, there are many tables, and each one needs to be rewritten. This makes maintenance more difficult in the future, which is a bit troublesome. I was worried about making mistakes, so I double-checked several times.

I initially thought that setting `enum Columns { static let name = Column(CodingKeys.name) }` would be enough. I hope a simpler method will be available in the future.

Thanks, If there's a simpler way, please be sure to let me know. Thank you so much.

**Comment by groue** [comment]
> I hope a simpler method will be available in the future.

In all honesty, that future might be very far away 😬 I do not quite see any solution in the current state of the language and its libraries.

- JSON serialization requires `Codable`, and Codable is not designed to easily switch keys between encoders. 🤷‍♂️ 
- GRDB could ship macros that mimic Codable features, but:
    - Macros have a [terrible compilation cost](https://forums.swift.org/t/compilation-extremely-slow-since-macros-adoption/67921) (only Apple can ship macros that do not eat _minutes_ when building a project). 🐌 
    - I'm not sure (euphemism) that macros can reproduce the features that people who are used to Codable can expect.
    - I'm very reluctant to force users to define records with macros. The current technique based on protocols allow users to _extend_ types in order to turn them into database records, and several users have expressed their appreciation of this feature.

**Comment by groue** [comment]
OK, so I close this issue. If you have more question, please chime in!

**Comment by SF-Simon** [comment]
> In all honesty, that future might be very far away 😬 I do not quite see any solution in the current state of the language and its libraries.

I understand this as well, because at the beginning we also considered trying to implement something similar to an alias in this way. However, we found it to be quite difficult, as it required converting and recognizing names for every process. 

Later, I had a change of perspective and decided to approach it from GRDB.swift. As a result, GRDB.swift didn’t disappoint me. Although it was a bit of a detour, it solved the problem. 

I realized that it doesn’t necessarily have to be a perfect solution。Sometimes saving some code can be a good option too.

Honestly, I feel like I can't do without GRDB.swift now. Thank you for bringing such a great project to us, Groue. I truly appreciate it.

**Comment by groue** [comment]
Thank you @SF-Simon, this is appreciated :-) The library tries to never lock the users out of their app, and I'm very pleased that it rarely disappoints. Now some solutions are more involved than others, and here we clearly are not in an ideal situation 😅

**Comment by SF-Simon** [comment]
> OK, so I close this issue. If you have more question, please chime in!

Sorry, we still have some questions regarding the usage mentioned earlier. We are using it with a class type, not a struct type. I noticed that in the source code, `init(row: row)` is marked with `throws`. Do we need to follow this rule? Also, when initializing, do we need to call `super.init()`?

```
    class Player: NSObject, FetchableRecord & PersistableRecord, Codable {
        var id: Int64
        var name: String
        var score: Int

        /// JSON Keys
        enum CodingKeys: String, CodingKey {
            case id
            case name = "title"
            case score
        }
        
        enum Columns {
            static let id = Column(CodingKeys.id)
            static let name = Column("name") // deviation from JSON
            static let score = Column(CodingKeys.score)
        }
        
        required init(row: Row) throws { // <----
            self.id = row[Columns.id]
            self.name = row[Columns.name]
            self.score = row[Columns.score]
            super.init() // <--
        }
        
        func encode(to container: inout PersistenceContainer) {
            container[Columns.id] = id
            container[Columns.name] = name
            container[Columns.score] = score
        }
            
    }
```
Currently, it seems that our usage is not causing any issues, but we're unsure if calling super.init() might result in it being executed multiple times. Just to be cautious, we’re reaching out to ask. If there are any considerations or changes we need to make to the initialization or encoding process due to the nature of the class, please let us know. 

Thank you.
